File: MiniDOMBase/DMD/Anagrafica/MotiviAppuntamento/MotivoAppuntamento.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases



Partial Public Class Anagrafica

    ''' <summary>
    ''' Flags per un Motivo appuntamento
    ''' </summary>
    Public Enum MotivoAppuntamentoFlags As Integer
        None = 0

        ''' <summary>
        ''' Attivo
        ''' </summary>
        Attivo = 1


        ''' <summary>
        ''' Valido per le persone fisiche
        ''' </summary>
        PersoneFisiche = 2

        ''' <summary>
        ''' Valido per le persone giuridiche
        ''' </summary>
        PersoneGiuridiche = 4
    End Enum

    ''' <summary>
    ''' Rappresenta un elemento nel menu di scelta a tendina per lo scopo degli appuntamenti
    ''' </summary>
    <Serializable>
    Public Class MotivoAppuntamento
        Inherits DBObject

        Private m_Nome As String
        Private m_Descrizione As String
        Private m_Flags As MotivoAppuntamentoFlags
        Private m_TipoAppuntamento As String

        Public Sub New()
            Me.m_Nome = ""
            Me.m_Descrizione = ""
            Me.m_Flags = MotivoAppuntamentoFlags.Attivo
            Me.m_TipoAppuntamento = ""
        End Sub

        ''' <summary>
        ''' Restituisce o imposta il nome della lista
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Nome As String
            Get
                Return Me.m_Nome
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_Nome
                If (oldValue = value) Then Exit Property
                Me.m_Nome = value
                Me.DoChanged("Nome", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta la descrizione della lista
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Descrizione As String
            Get
                Return Me.m_Descrizione
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_Descrizione
                If (oldValue = value) Then Exit Property
                Me.m_Descrizione = value
                Me.DoChanged("Descrizione", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta il nome del proprietario
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property TipoAppuntamento As String
            Get
                Return Me.m_TipoAppuntamento
            End Get
            Set(value As String)
                value = Strings.Trim(value)
                Dim oldValue As String = Me.m_TipoAppuntamento
                If (oldValue = value) Then Exit Property
                Me.m_TipoAppuntamento = value
                Me.DoChanged("TipoAppuntamento", value, oldValue)
            End Set
        End Property

        Public Property Flags As MotivoAppuntamentoFlags
            Get
                Return Me.m_Flags
            End Get
            Set(value As MotivoAppuntamentoFlags)
                Dim oldValue As MotivoAppuntamentoFlags = Me.m_Flags
                If (oldValue = value) Then Return
                Me.m_Flags = value
                Me.DoChanged("Flags", value, oldValue)
            End Set
        End Property

        Public Overrides Function ToString() As String
            Return Me.m_Nome
        End Function

        Public Overrides Function GetModule() As CModule
            Return Anagrafica.MotiviAppuntamento.Module
        End Function


        Public Overrides Function GetTableName() As String
            Return "tbl_CRMMotiviAppuntamento"
        End Function

        Protected Overrides Function LoadFromRecordset(reader As DBReader) As Boolean
            Me.m_Nome = reader.Read("Nome", Me.m_Nome)
            Me.m_Descrizione = reader.Read("Descrizione", Me.m_Descrizione)
            Me.m_Flags = reader.Read("Flags", Me.m_Flags)
            Me.m_TipoAppuntamento = reader.Read("TipoAppuntamento", Me.m_TipoAppuntamento)
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Protected Overrides Function SaveToRecordset(writer As DBWriter) As Boolean
            writer.Write("Nome", Me.m_Nome)
            writer.Write("Descrizione", Me.m_Descrizione)
            writer.Write("Flags", Me.m_Flags)
            writer.Write("TipoAppuntamento", Me.m_TipoAppuntamento)
            Return MyBase.SaveToRecordset(writer)
        End Function

        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            writer.WriteAttribute("Nome", Me.m_Nome)
            writer.WriteAttribute("Flags", Me.m_Flags)
            writer.WriteAttribute("TipoAppuntamento", Me.m_TipoAppuntamento)
            MyBase.XMLSerialize(writer)
            writer.WriteTag("Descrizione", Me.m_Descrizione)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldNome As String, fieldValue As Object)
            Select Case fieldNome
                Case "Nome" : Me.m_Nome = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Descrizione" : Me.m_Descrizione = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Flags" : Me.m_Flags = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "TipoAppuntamento" : Me.m_TipoAppuntamento = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldNome, fieldValue)
            End Select
        End Sub

        Protected Overrides Sub OnAfterSave(e As SystemEvent)
            MyBase.OnAfterSave(e)
            Anagrafica.MotiviAppuntamento.UpdateCached(Me)
        End Sub

        Protected Friend Overrides Function GetConnection() As CDBConnection
            Return Databases.APPConn
        End Function
    End Class


End Class

File: MiniDOMBase/DMD/Anagrafica/MotiviAppuntamento/MotivoAppuntamento.cs
using System;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using static minidom.Sistema;
using minidom.repositories;

namespace minidom
{
    public partial class Anagrafica
    {

        /// <summary>
        /// Flags per un Motivo appuntamento
        /// </summary>
        [Flags]
        public enum MotivoAppuntamentoFlags : int
        {
            /// <summary>
            /// Nessun flag
            /// </summary>
            None = 0,

            /// <summary>
            /// Attivo
            /// </summary>
            Attivo = 1,


            /// <summary>
            /// Valido per le persone fisiche
            /// </summary>
            PersoneFisiche = 2,

            /// <summary>
            /// Valido per le persone giuridiche
            /// </summary>
            PersoneGiuridiche = 4
        }

        /// <summary>
        /// Rappresenta un elemento nel menu di scelta a tendina per lo scopo degli appuntamenti
        /// </summary>
        [Serializable]
        public class MotivoAppuntamento 
            : Databases.DBObject
        {
            private string m_Nome;
            private string m_Descrizione;
            private string m_TipoAppuntamento;
            
            /// <summary>
            /// Costruttore
            /// </summary>
            public MotivoAppuntamento()
            {
                this.m_Nome = "";
                this.m_Descrizione = "";
                this.m_Flags = (int)MotivoAppuntamentoFlags.Attivo;
                this.m_TipoAppuntamento = "";
            }

             
            /// <summary>
            /// Restituisce o imposta il nome della lista
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string Nome
            {
                get
                {
                    return m_Nome;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_Nome;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Nome = value;
                    DoChanged("Nome", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta la descrizione della lista
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string Descrizione
            {
                get
                {
                    return m_Descrizione;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_Descrizione;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Descrizione = value;
                    DoChanged("Descrizione", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta il nome del proprietario
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string TipoAppuntamento
            {
                get
                {
                    return m_TipoAppuntamento;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_TipoAppuntamento;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_TipoAppuntamento = value;
                    DoChanged("TipoAppuntamento", value, oldValue);
                }
            }

            /// <summary>
            /// Flags
            /// </summary>
            public new MotivoAppuntamentoFlags Flags
            {
                get
                {
                    return (MotivoAppuntamentoFlags) this.m_Flags;
                }

                set
                {
                    var oldValue = this.Flags;
                    if (oldValue == value)
                        return;
                    this.m_Flags = (int)value;
                    DoChanged("Flags", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce una stringa che rappresenta l'oggetto
            /// </summary>
            /// <returns></returns>
            public override string ToString()
            {
                return m_Nome;
            }

            /// <summary>
            /// Restituisce il codice hash dell'oggetto
            /// </summary>
            /// <returns></returns>
            public override int GetHashCode()
            {
                return HashCalculator.Calculate(this.m_Nome);
            }

            /// <summary>
            /// Restituisce true se gli oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public sealed override bool Equals(Databases.DBObject obj)
            {
                return (obj is MotivoAppuntamento) && this.Equals((MotivoAppuntamento)obj);
            }

            /// <summary>
            /// Restituisce true se gli oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public virtual bool Equals(MotivoAppuntamento obj)
            {
                return base.Equals(obj)
                     && DMD.Strings.EQ(this.m_Nome, obj.m_Nome)
                     && DMD.Strings.EQ(this.m_Descrizione, obj.m_Descrizione)
                     && DMD.Strings.EQ(this.m_TipoAppuntamento, obj.m_TipoAppuntamento)
                    ;
            //private CKeyCollection m_Parameters;
            }



            /// <summary>
            /// Repository
            /// </summary>
            /// <returns></returns>
            public override CModulesClass GetModule()
            {
                return minidom.Anagrafica.MotiviAppuntamento; //.Module;
            }

            /// <summary>
            /// Discriminator
            /// </summary>
            /// <returns></returns>
            public override string GetTableName()
            {
                return "tbl_CRMMotiviAppuntamento";
            }

            /// <summary>
            /// Carica dal db
            /// </summary>
            /// <param name="reader"></param>
            /// <returns></returns>
            protected override bool LoadFromRecordset(DBReader reader)
            {
                this.m_Nome = reader.Read("Nome", this.m_Nome);
                this.m_Descrizione = reader.Read("Descrizione", this.m_Descrizione);
                this.m_TipoAppuntamento = reader.Read("TipoAppuntamento", this.m_TipoAppuntamento);
                return base.LoadFromRecordset(reader);
            }

            /// <summary>
            /// Salva nel db
            /// </summary>
            /// <param name="writer"></param>
            /// <returns></returns>
            protected override bool SaveToRecordset(DBWriter writer)
            {
                writer.Write("Nome", m_Nome);
                writer.Write("Descrizione", m_Descrizione);
                writer.Write("TipoAppuntamento", m_TipoAppuntamento);
                return base.SaveToRecordset(writer);
            }

            /// <summary>
            /// Prepara i campi
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaFields(DBTable table)
            {
                base.PrepareDBSchemaFields(table);

                var c = table.Fields.Ensure("Nome", typeof(string), 255);
                c = table.Fields.Ensure("Descrizione", typeof(string), 0);
                c = table.Fields.Ensure("TipoAppuntamento", typeof(string), 255);
            }

            /// <summary>
            /// Prepara i vincoli
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaConstraints(DBTable table)
            {
                base.PrepareDBSchemaConstraints(table);

                var c = table.Constraints.Ensure("idxNome", new string[] { "Nome" }, DBFieldConstraintFlags.None);
                c = table.Constraints.Ensure("idxTipo", new string[] { "TipoAppuntamento" }, DBFieldConstraintFlags.None);
                c = table.Constraints.Ensure("idxDescrizione", new string[] { "Descrizione" }, DBFieldConstraintFlags.None);
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("Nome", m_Nome);
                writer.WriteAttribute("TipoAppuntamento", m_TipoAppuntamento);
                base.XMLSerialize(writer);
                writer.WriteTag("Descrizione", m_Descrizione);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldNome"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldNome, object fieldValue)
            {
                switch (fieldNome ?? "")
                {
                    case "Nome":
                        {
                            m_Nome = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Descrizione":
                        {
                            m_Descrizione = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "TipoAppuntamento":
                        {
                            m_TipoAppuntamento = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }
                     
                    default:
                        {
                            base.SetFieldInternal(fieldNome, fieldValue);
                            break;
                        }
                }
            }

            //protected override void OnAfterSave(SystemEvent e)
            //{
            //    base.OnAfterSave(e);
            //    MotiviAppuntamento.UpdateCached(this);
            //}

            //protected internal override Databases.CDBConnection GetConnection()
            //{
            //    return Databases.APPConn;
            //}
        }
    }
}